### PR TITLE
Daubechies Wavelets

### PR DESCRIPTION
**Pull request by sjperkins**
Still a WIP

### CODE REVIEW

**Comment by sjperkins** [comment]
Thanks @landmanbester :duck: :duck: :duck:

**Comment by landmanbester** [comment]
Oh, nice! Glad to be a good little rubber duck ;-)

**Comment by sjperkins** [comment]
@landmanbester Could you please review?

**Comment by sjperkins** [comment]
I've made an effort to reproduce the pywavelets interfaces on the provided functions.

On the reconstruction side of things, pywavelets supports missing approximation/detail coefficients. This PR doesn't support this, but with some type mangling, it may be possible if it's needed. I can only imagine this happening if coefficients got deleted from the dictionary produced by wavedecn call and then fed into a waverecn call.

**Comment by landmanbester** [comment]
Cool, thanks. I'll have a look asap

**Comment by landmanbester** [comment]
> I've made an effort to reproduce the pywavelets interfaces on the provided functions.
> 
> On the reconstruction side of things, pywavelets supports missing approximation/detail coefficients. This PR doesn't support this, but with some type mangling, it may be possible if it's needed. I can only imagine this happening if coefficients got deleted from the dictionary produced by wavedecn call and then fed into a waverecn call.

I don't see this being an issue

File: pfb/test/test_wavelets.py
Comment (landmanbester) [inline pfb/test/test_wavelets.py:165]: Can we parametrise over the number of levels here as well please?

File: pfb/test/test_wavelets.py
Comment (landmanbester) [inline pfb/test/test_wavelets.py:225]: Can we test this with multiple levels as well please?

File: pfb/test/test_wavelets.py
Comment (landmanbester) [inline pfb/test/test_wavelets.py:241]: I usually use this with mode='periodization' but I see below that it has not been implemented. I don't know how big a difference this makes. Will have to test

File: pfb/wavelets/convolution.py
Comment (landmanbester) [inline pfb/wavelets/convolution.py:89]: See comment above I think I need mode='periodic' support? How hard would it be to add this?

**Review by landmanbester — CHANGES_REQUESTED** [review]
Thanks @sjperkins this looks good and it seems like it was quite a bit of work. The only thing I would need is to make sure that we get consistent results with varying decomposition levels. We can probably test this by parametrising over level in the waverecn test. I need to follow up on the importance of the mode='periodisation'  argument. I'll do some tests once this is merged and also check with Audrey

File: pfb/wavelets/convolution.py
Comment (sjperkins) [reply pfb/wavelets/convolution.py:89]: Unimplemented modes need their loops filled in from [here](https://github.com/PyWavelets/pywt/blob/0136844b47d318697c05a8f69ae3755540532a84/pywt/_extensions/c/convolution.template.c#L127-L470).

I think zeropad is implemented (although I haven't tested parity)

Comment (sjperkins) [reply pfb/wavelets/convolution.py:89]: Note that periodisation is actually implemented in a separate function but it might no longer be strictly required based on email conversations?

File: pfb/wavelets/convolution.py
Comment (sjperkins) [inline pfb/wavelets/convolution.py:45]: Periodisation is actually handled here and should defer to a function specifically for it.

File: pfb/test/test_wavelets.py
Comment (sjperkins) [reply pfb/test/test_wavelets.py:241]: Would it not be possible to use zeropad, in light of email conversations?

File: pfb/test/test_wavelets.py
Comment (sjperkins) [reply pfb/test/test_wavelets.py:225]: Yes, it should be possible to parametrize this on level.

**Comment by sjperkins** [comment]
> Thanks @sjperkins this looks good and it seems like it was quite a bit of work. The only thing I would need is to make sure that we get consistent results with varying decomposition levels. We can probably test this by parametrising over level in the waverecn test. I need to follow up on the importance of the mode='periodisation' argument. I'll do some tests once this is merged and also check with Audrey

Parametrising the decomposition tests over level should by easy enough

I suspect we can just use the zeropad mode now. This probably needs a test.

**Comment by landmanbester** [comment]
Yep, we just need zeropad mode. This should be the simplest so hopefully not too much work to add it in?

**Comment by sjperkins** [comment]
> Yep, we just need zeropad mode. This should be the simplest so hopefully not too much work to add it in?

I think it might be done -- I put the code paths in. I just haven't tested the mode yet.

**Comment by sjperkins** [comment]
@landmanbester I've parametrised the wavedecn/waverecn tests over

1. 10 levels
2. real and complex data
3. 3 daubechies wavelets
4. symmetric and zero padding modes.

All the test cases pass. Note that in the case of (1), the number of levels can be too high. pywavelets warns, but in numba this would require dropping to [objmode](https://numba.pydata.org/numba-doc/latest/user/withobjmode.html) to log the warning. It looks like it may be possible with bleeding edge numba (https://github.com/numba/numba/pull/5674) but I've commented that out for now, since the wavedecn/waverecn compiles take a while and I'd prefer the compilations to be cached.

**Comment by landmanbester** [comment]
Awesome! I'm gonna merge and test how it parallelises. Thanks @sjperkins for this large and valuable contribution

**Pull request merged**
